File: consumer/src/main/java/boojongmin/bank/consumer/factory/Factory.java
package boojongmin.bank.consumer.factory;

import boojongmin.bank.consumer.property.BankProperties;
import org.apache.kafka.clients.consumer.KafkaConsumer;
import org.apache.kafka.clients.producer.KafkaProducer;

public class Factory {
	public static KafkaConsumer<String, String> createConsumer(String topic) {
		return new KafkaConsumer<>(BankProperties.createConsumerProperties(topic));
	}

	public static KafkaProducer<String, String> createProducer() {
		return new KafkaProducer<>(BankProperties.createProducerProperties());
	}
}

File: common/src/main/kotlin/boojongmin/bank/Factory.kt
package boojongmin.bank

import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.databind.SerializationFeature
import com.fasterxml.jackson.module.kotlin.KotlinModule
import org.apache.kafka.clients.consumer.KafkaConsumer
import org.apache.kafka.clients.producer.KafkaProducer
import org.apache.kafka.common.TopicPartition
import java.util.*

object Factory {
    val properties = Properties()
    val clazzLoader = javaClass.classLoader

    fun createConsumer(partition: Int): KafkaConsumer<String, String> {
        properties.load(clazzLoader.getResourceAsStream("consumer.properties"))
        val consumer = KafkaConsumer<String, String>(properties)
        val topicPartition = LogStep.values().map {TopicPartition(it.name, partition)}
        consumer.assign(topicPartition)
        return consumer
    }

    fun createProducer(): KafkaProducer<String, String> {
        properties.load(clazzLoader.getResourceAsStream("producer.properties"))
        return KafkaProducer(properties)
    }

    fun createObjectMapper(): ObjectMapper {
        val mapper = ObjectMapper().registerModule(KotlinModule())
        mapper.enable(SerializationFeature.INDENT_OUTPUT)
        return mapper
    }
}

